web-ng/src/app/services/router/router.service.ts

```diff
@@ -52,7 +58,17 @@ export class RouterService {
     this.activatedRoute = route;
     // Pipe values from URL query parameters.
     this.projectId$ = route.paramMap.pipe(
-      map(params => params.get('projectId'))
+      map(params => {
+        if (params.get('projectId') === ':new') {
+          const projectId = this.dataStoreService.generateId();
+          this.projectService.createProject(projectId).then(() => {
```

Comment:
In functional programming, side-effects (functions, or in this case streams, that modifying state) are generally avoided since they break "purity", one of the key qualities of the paradigm, including safe multi-threading and easy debuggability.

Also, following the Single Responsibility Principle, the RouterService's one responsibility is to abstract the URL state. Saving the project is out of scope.

To fix this, the main page component should subscribe to the `projectId$` stream, and when the id is `:new`, it should create a new project there instead.

Reply:
Actually, this isn't the expected behavior; as described in #313, we'd expect the project to be created in the db only when the title is saved. The same would apply for saving layers and sharing; the project would get created on the first change an no sooner. This prevents the db from getting filled with empty abandoned projects. I'll look at the code now to see where the hook points might be.

Reply:
I think it's ok the redirect inside `ProjectService` or `LayerService` once a new project has been stored to the db, i.e.:

* `updateTitle` in `ProjectService` generates new id if necessary, saves, and calls `RouterService.setProjectId(..)` to new id.
* `updateLayer` in `LayerService` (#369) does the same.

Wdyt?

Reply:
Thanks I will work on them soon :)